src/ci/stage-build.py

```diff
@@ -754,95 +756,121 @@ def record_metrics(pipeline: Pipeline, timer: Timer):
     if metrics is None:
         return
     llvm_steps = tuple(metrics.find_all_by_type("bootstrap::llvm::Llvm"))
-    assert len(llvm_steps) > 0
     llvm_duration = sum(step.duration for step in llvm_steps)
 
     rustc_steps = tuple(metrics.find_all_by_type("bootstrap::compile::Rustc"))
-    assert len(rustc_steps) > 0
     rustc_duration = sum(step.duration for step in rustc_steps)
 
     # The LLVM step is part of the Rustc step
-    rustc_duration -= llvm_duration
+    rustc_duration = max(0, rustc_duration - llvm_duration)
 
-    timer.add_duration("LLVM", llvm_duration)
-    timer.add_duration("Rustc", rustc_duration)
+    if llvm_duration > 0:
+        timer.add_duration("LLVM", llvm_duration)
+    if rustc_duration > 0:
+        timer.add_duration("Rustc", rustc_duration)
 
     log_metrics(metrics)
 
 
-def execute_build_pipeline(timer: Timer, pipeline: Pipeline, runner: BenchmarkRunner, final_build_args: List[str]):
+def execute_build_pipeline(timer: Timer, pipeline: Pipeline, runner: BenchmarkRunner, dist_build_args: List[str]):
     # Clear and prepare tmp directory
     shutil.rmtree(pipeline.opt_artifacts(), ignore_errors=True)
     os.makedirs(pipeline.opt_artifacts(), exist_ok=True)
 
     pipeline.build_rustc_perf()
 
-    # Stage 1: Build rustc + PGO instrumented LLVM
-    with timer.section("Stage 1 (LLVM PGO)") as stage1:
-        with stage1.section("Build rustc and LLVM") as rustc_build:
-            build_rustc(pipeline, args=[
-                "--llvm-profile-generate"
-            ], env=dict(
-                LLVM_PROFILE_DIR=str(pipeline.llvm_profile_dir_root() / "prof-%p")
-            ))
+    """
+    Stage 1: Build PGO instrumented rustc
+
+    We use a normal build of LLVM, because gathering PGO profiles for LLVM and `rustc` at the same time
+    can cause issues.
```

Comment:
What sort of issues does it cause? i guess it makes the overall build take longer?

Reply:
Historically the problem was incompatibility of the profiling/instrumentation machinery. When we PGO instrument LLVM, we do that using the host LLVM. But when we PGO instrument `rustc`, we do that using the in-tree LLVM. There are not always the same versions (sometimes also the major version is different!), and the two instrumentation runtimes then collide. At least that is my understanding of what was happening. I can try this scenario again to see what would happen.

Reply:
hmm, ok. why do we use different versions of LLVM between the two? i think if we consistently used the host LLVM, we could avoid a rebuild of "normal" in-tree LLVM (the one you said in the PR description that you wanted to cache)

Reply:
Hmm. I always thought that we have to use our patched LLVM in order to build `rustc` (and in order to build other Rust programs), is that not so?

If not, then in theory we could only use the host LLVM, although we would have to convince bootstrap to actually use it. I'm not sure if the host LLVM can handle everything that we need though, so we would maybe need to recompile it anyway, and in that case it's probably not better than just using the in-tree LLVM.

Reply:
When you say "host llvm" I'm assuming you mean `llvm-config = ...` in config.toml. We support older versions of LLVM for *building* (currently llvm 14), but I don't see any reason we need to support them for *PGO* - I think it's fine to give a hard error if the host version of LLVM is too low.

If `llvm-config` *isn't* set, then we're using in-tree LLVM or download-ci-llvm anyway, and we know that it's the right version and the instrumentation will work.

Reply:
I don't think that we use `llvm-config` in `config.toml`, because in that case LLVM wouldn't get built at all, right? We use the host LLVM (clang) to compile in-tree LLVM with instrumentation.

Reply:
Dist builds cannot or shouldn't use host LLVM for building rustc, as that's not tested as widely and may be slower (as it lacks some custom patches, or at least did historically)